### PR TITLE
[FEA] Type inferred unaryops

### PR DESCRIPTION
**Pull request by mzegar**
`CUDF` actually throws if the `df.cos()` receives a Series that isn't of type numeric.
```python
import cudf
df = cudf.DataFrame({'a': [3, 2, 3, 4], 'b': [4, 3, 2, 4], 'c': ["foo", "bar", "foo", "bar"]})
print(df.cos()) # error
```

This change now makes it so that if any of the `Series` inside of the `DataFrame` are not of type `Numeric`, the return type is `never`. Previously, we would just skip over and ignore any `Series` that isn't type `Numeric`. So this change brings it closer in line with `CUDF`.

### CODE REVIEW

**Comment by AjayThorve** [comment]
definitely prefer this way over the current version. Also, not sure if we skipped over non-numeric series in the current version, as it seems to include the StringSeries in the return hints as well

image

**Comment by mzegar** [comment]
> definitely prefer this way over the current version. Also, not sure if we skipped over non-numeric series in the current version, as it seems to include the StringSeries in the return hints as well
> 
> image

Yeah we skipped over any non numeric series and kept it in the dataframe (in the current implemention)

File: modules/cudf/src/data_frame.ts [hunk 1]
```diff
@@ -1022,16 +1015,10 @@ export class DataFrame<T extends TypeMap = any> {
...
-        ...map,
-        [name]: _invokeIfNumericSeries(
-          this.get(name),
-          () => { return Series.new(this._accessor.get(name).sin(memoryResource)); })
-      };
-    };
-    return new DataFrame(this.names.reduce(mapper, {} as SeriesMap<T>));
+  sin<P extends keyof T>(memoryResource?: MemoryResource) {
+    return new DataFrame(this.names.reduce(
+             (map, name) => ({...map, [name]: (this.get(name) as any).sin(memoryResource)}),
```
Comment (trxcllnt) [inline modules/cudf/src/data_frame.ts:1020]: Should we throw a more descriptive error here if the types are wrong? Right now it'll throw something like `TypeError: this.get(...).sin is not a function`

**Review by trxcllnt — CHANGES_REQUESTED** [review]
Probably good to add more descriptive error messages at runtime.

File: modules/cudf/src/data_frame.ts (hunk 1 above)
Comment (mzegar) [reply modules/cudf/src/data_frame.ts]: Spoke with Paul, I assumed that `typescript` would be a requirement to use `node-rapids`, but this isn't the case.

Regardless, this PR should provide better typing if people do decide to use typescript (i.e return `never` if the result won't occur), but also giving useful `throw` messages as well.

File: modules/cudf/src/data_frame.ts [hunk 2]
```diff
@@ -71,16 +77,43 @@ type JoinProps<
...
+  new Int16,
+  new Int32,
+  new Int64,
+  new Uint8,
+  new Uint16,
+  new Uint32,
+  new Uint64,
+];
+
+const FloatTypes = [new Float32, new Float64];
```
Comment (mzegar) [inline modules/cudf/src/data_frame.ts:105]: Do we maybe want to move these to `dtypes.ts` and export them?

Comment (trxcllnt) [reply modules/cudf/src/data_frame.ts:105]: Yeah I think that'd be good.

**Review by trxcllnt — APPROVED** [review]
LGTM. I don't think we should block this on moving the dtypes, but feel free if you want to.

**Comment by mzegar** [comment]
I'll create an issue for this.

**Pull request merged**